### PR TITLE
Sp3 changes + unittesting

### PR DESCRIPTION
**Pull request by seballgeyer**
Set of changes in the read SP3 triggered by trying to read a corrupted sp3 file. now each epoch is read separately to be able to read most of a corrupted file. 
conversion / add of some of the preexisting test into the python unittest package framework & creation of a pipeline to run the test at the push on the repo. 

More to come.

### CODE REVIEW

File: .github/workflows/python-cicd-units.yaml
Comment (ronaldmaj) [inline .github/workflows/python-cicd-units.yaml:1]: Cool, appreciate you putting this in, it'll help keep the code base functional with new changes

We'll need to expand the unit testing further to cover more of the existing functions, but that will come with time

File: gnssanalysis/gn_io/sp3.py [hunk 1]
```diff
@@ -59,12 +54,10 @@ def sp3_pos_nodata_to_nan(
         & (sp3_df[("EST", "Y")] == SP3_POS_NODATA_NUMERIC)
         & (sp3_df[("EST", "Z")] == SP3_POS_NODATA_NUMERIC)
     )
-    sp3_df.loc[nan_mask, [("EST", "X"), ("EST", "Y"), ("EST", "Z")]] = _np.NAN
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:62]: As I understand `np.nan` `np.NaN` `np.NAN` are all equivalent? If that's the case, we'll just need to choose one and be consistent. Happy to go with `_np.nan` unless there are clear objections from others / an argument to be made that this choice is detrimental somehow

File: gnssanalysis/gn_io/sp3.py [hunk 2]
```diff
@@ -44,9 +41,7 @@
 SP3_POS_STD_NODATA = -100
 
 
-def sp3_pos_nodata_to_nan(
-    sp3_df: _pd.DataFrame
-) -> None:
+def sp3_pos_nodata_to_nan(sp3_df: _pd.DataFrame) -> None:
     """
     Converts the SP3 Positional column's nodata values (0.000000) to NaNs.
     See https://files.igs.org/pub/data/format/sp3_docu.txt
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:47]: Just on the path of filling in docstrings for older functions, it might be a good idea that for all functions that we touch in future PRs, we should aim to complete the one-line sphinx docstring.
We can either start with your PR here, or for all future ones

File: gnssanalysis/gn_io/sp3.py [hunk 3]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
     return off, scl
 
 
+def _process_sp3_block(date, data, widths, names):
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:81]: Type-hints missing

File: gnssanalysis/gn_io/sp3.py [hunk 4]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
...
+    """Process a single block of SP3 data"""
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:82]: Should be a one-line sphinx style docstring

File: gnssanalysis/gn_io/sp3.py [hunk 5]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
...
+    temp_sp3.set_index(temp_sp3.PRN.astype(str), append=True, inplace=True)
+    temp_sp3.set_index(temp_sp3.PV_FLAG.astype(str), append=True, inplace=True)
+    return temp_sp3
+
+
 def read_sp3(sp3_path, pOnly=True, nodata_to_nan=True):
     """Read SP3 file
     Returns STD values converted to proper units (mm/ps) also if present.
     by default leaves only P* values (positions), removing the P key itself
     """
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:99]: Again, could be good to complete docstring here

File: gnssanalysis/gn_io/sp3.py [hunk 6]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
...
+    # rename columsn x_coordinate -> [EST, X], y_coordinate -> [EST, Y]
+    sp3_df.columns = [
+        ["EST", "EST", "EST", "EST", "STD", "STD", "STD", "STD", "a1", "a2", "a3", "a4"],
+        ["X", "Y", "Z", "CLK", "X", "Y", "Z", "CLK", "", "", "", ""],
+    ]
     if nodata_to_nan:
         sp3_pos_nodata_to_nan(sp3_df)  # Convert 0.000000 (which indicates nodata in the SP3 POS column) to NaN
         sp3_clock_nodata_to_nan(sp3_df)  # Convert 999999* (which indicates nodata in the SP3 CLK column) to NaN
 
+    # print(sp3_df.index.has_duplicates())
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:156]: Can probably delete this comment?

File: gnssanalysis/gn_io/sp3.py [hunk 7]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
...
+        "c_sdev",
+        "Unused5",
+        "Clock_Event_Flag",
+        "Clock_Pred_Flag",
+        "Unused6",
+        "Maneuver_Flag",
+        "Orbit_Pred_Flag",
+    ]
+    name_float = ["x_coordinate", "y_coordinate", "z_coordinate", "clock", "x_sdev", "y_sdev", "z_sdev", "c_sdev"]
+    sp3_df = _pd.concat([_process_sp3_block(date, data, widths, names) for date, data in zip(date_lines, data_blocks)])
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:142]: So this `_process_sp3_block` function becomes the core of the parsing process?

File: gnssanalysis/gn_io/sp3.py [hunk 8]
```diff
@@ -228,7 +204,7 @@ def parse_sp3_header(header):
 def getVelSpline(sp3Df):
     """returns in the same units as intput, e.g. km/s (needs to be x10000 to be in cm as per sp3 standard"""
     sp3dfECI = sp3Df.EST.unstack(1)[["X", "Y", "Z"]]  # _ecef2eci(sp3df)
-    datetime = sp3dfECI.index.values
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:231]: Did this not work previously?

File: gnssanalysis/gn_io/sp3.py [hunk 9]
```diff
@@ -228,7 +204,7 @@ def parse_sp3_header(header):
 def getVelSpline(sp3Df):
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:204]: Type-hints / docstring

File: gnssanalysis/gn_io/sp3.py [hunk 10]
```diff
@@ -496,7 +472,7 @@ def sp3merge(sp3paths, clkpaths=None, nodata_to_nan=False):
     return merged_sp3
 
 
-def sp3_hlm_trans(a: _pd.DataFrame, b: _pd.DataFrame) -> tuple((_pd.DataFrame, list)):
+def sp3_hlm_trans(a: _pd.DataFrame, b: _pd.DataFrame) -> tuple[_pd.DataFrame, list]:
     """Rotates sp3_b into sp3_a. Returns a tuple of updated sp3_b and HLM array with applied computed parameters and residuals"""
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:476]: One-line sphinx docstring

File: gnssanalysis/gn_transform.py [hunk 11]
```diff
@@ -25,20 +26,18 @@ def gen_helm_aux(pt1, pt2):
...
-        [unity_blk, xyz_blk.reshape((-1, 3)), xyz]
-    ) 
+    A = _np.column_stack([unity_blk, xyz_blk.reshape((-1, 3)), xyz])
     rhs = pt2.reshape((-1, 1)) - xyz  # right-hand side
     return A, rhs
 
 
-def get_helmert7(pt1:_np.ndarray, pt2:_np.ndarray, scale_in_ppm:bool = True):
+def get_helmert7(pt1: _np.ndarray, pt2: _np.ndarray, scale_in_ppm: bool = True):
     """inversion of 7 Helmert parameters between 2 sets of points. pt1@hlm -> pt2"""
```
Comment (ronaldmaj) [inline gnssanalysis/gn_transform.py:35]: Complete docstring (one-line sphinx)

File: gnssanalysis/gn_transform.py [hunk 12]
```diff
@@ -55,7 +54,7 @@ def gen_rot_matrix(v):
     return mat + _np.eye(3)
 
 
-def transform7(xyz_in, hlm_params, scale_in_ppm:bool = True):
+def transform7(xyz_in, hlm_params, scale_in_ppm: bool = True):
```
Comment (ronaldmaj) [inline gnssanalysis/gn_transform.py:57]: Complete type-hints

File: gnssanalysis/gn_transform.py [hunk 13]
```diff
@@ -55,7 +54,7 @@ def gen_rot_matrix(v):
...
     """transformation of xyz vector with 7 helmert parameters. The helmert parameters array consists of the following:
     Three transformations Tx, Ty, Tz usually in meters, or the coordinate units used for the computation of the hlm parameters.
     Three rotations Rx, Ry and Rz in radians.
```
Comment (ronaldmaj) [inline gnssanalysis/gn_transform.py:60]: Complete sphinx style docstring

File: tests/test_sp3.py
Comment (ronaldmaj) [inline tests/test_sp3.py:1]: I haven't run a test of this myself, but looks good from glancing through it.
Would be good to get @EugeneDu-GA to look at his sp3 files and see if there are any other corner cases that we should add here

File: gnssanalysis/gn_io/sp3.py [hunk 14]
```diff
@@ -85,104 +78,86 @@ def mapparm(old, new):
...
+    temp_sp3 = _pd.read_fwf(_io.StringIO(data), widths=widths, names=names)
+    dt_index = _np.repeat(a=_gn_datetime.datetime2j2000(epochs_dt.values), repeats=len(temp_sp3))
+    temp_sp3.set_index(dt_index, inplace=True)
+    temp_sp3.index.name = "J2000"
+    temp_sp3.set_index(temp_sp3.PRN.astype(str), append=True, inplace=True)
+    temp_sp3.set_index(temp_sp3.PV_FLAG.astype(str), append=True, inplace=True)
+    return temp_sp3
+
+
 def read_sp3(sp3_path, pOnly=True, nodata_to_nan=True):
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:95]: The output type-hint is missing here. 
Also, I think @EugeneDu-GA should test any sp3 files he was having trouble with against this function

**Review by ronaldmaj — CHANGES_REQUESTED** [review]
This is a big contribution and cleans up a lot of the code, thanks a lot for this!
The testing you've introduced is fantastic and I think it's definitely something all of us will need to continue with future PRs.

File: gnssanalysis/gn_io/sp3.py [hunk 15]
```diff
@@ -73,116 +91,86 @@ def sp3_clock_nodata_to_nan(
     :return None
     """
     nan_mask = sp3_df[("EST", "CLK")] >= SP3_CLOCK_NODATA_NUMERIC
-    sp3_df.loc[nan_mask, ("EST", "CLK")] = _np.NAN
+    sp3_df.loc[nan_mask, ("EST", "CLK")] = _np.nan
+
 
+def mapparm(old: Tuple[float, float], new: Tuple[float, float]) -> Tuple[float, float]:
+    """
+    Maps values from the old range to the new range.
```
Comment (treefern) [inline gnssanalysis/gn_io/sp3.py:99]: Would it be more accurate for this docstring to say this _generates_ scale and offset (mapping parameters) **to enable** remapping. This function isn't for direct data processing.

File: gnssanalysis/gn_io/sp3.py [hunk 16]
```diff
@@ -35,6 +32,31 @@
 # File descriptor and clock
 _RE_SP3_HEAD_FDESCR = _re.compile(rb"\%c[ ]+(\w{1})[ ]+cc[ ](\w{3})")
 
+
+_SP3_DEF_PV_WIDTH = [1, 3, 14, 14, 14, 14, 1, 2, 1, 2, 1, 2, 1, 3, 1, 1, 1, 1, 1, 1]
```
Comment (treefern) [inline gnssanalysis/gn_io/sp3.py:36]: Good stuff. Thanks for this, I'd been meaning to make the use of FWF more robust.

File: .github/workflows/python-cicd-units.yaml [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install boto3 click hatanaka matplotlib numpy pandas plotext==4.2 plotly pymongo pytest scipy tqdm unlzw3 typing_extensions
```
Comment (treefern) [inline .github/workflows/python-cicd-units.yaml:18]: This duplicates specification of our requirements. It's not as simple as just pointing at `requirements.txt` though, as there isn't one.

Requirements are specified in `setup.py`, though that _could_ be modified to read them from a `requirements.txt` file for consistency... I prototyped that idea and it works, but it's hacky because I have to handle possible commenting. Time for me to read about the differences between setup.py and requirements.txt and how to do this properly...

File: .github/workflows/python-cicd-units.yaml [hunk 18]
```diff
@@ -0,0 +1,21 @@
...
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up Python 3.10
+        uses: actions/setup-python@v2
+        with:
+          python-version: "3.10"
+      - name: Install dependencies
+        run: |
+          python -m pip install --upgrade pip
+          pip install  -r requirements.txt
```
Comment (treefern) [inline .github/workflows/python-cicd-units.yaml:18]: After looking at it a bit more I think there's a better way than my previous suggestion.

I'm dropping the suggestion in line here, based on my comment on the [specific commit](https://github.com/GeoscienceAustralia/gnssanalysis/commit/a8bcf609832303fb2f66ebc8bf139950aaef8852#r144575145)
```suggestion
          sed -i "/{{VERSION_PLACEHOLDER}}/d" setup.py
          pip install .
```

File: gnssanalysis/gn_io/sp3.py [hunk 19]
```diff
@@ -73,116 +91,86 @@ def sp3_clock_nodata_to_nan(
...
+    return temp_sp3
+
+
+def read_sp3(sp3_path: str, pOnly: bool = True, nodata_to_nan: bool = True) -> _pd.DataFrame:
+    """Reads an SP3 file and returns the data as a pandas DataFrame.
+
+
+    :param str sp3_path: The path to the SP3 file.
+    :param bool pOnly: If True, only P* values (positions) are included in the DataFrame. Defaults to True.
+    :param book nodata_to_nan: If True, converts 0.000000 (indicating nodata) to NaN in the SP3 POS column
```
Comment (treefern) [inline gnssanalysis/gn_io/sp3.py:141]: Typo
```suggestion
    :param bool nodata_to_nan: If True, converts 0.000000 (indicating nodata) to NaN in the SP3 POS column
```

File: gnssanalysis/gn_io/sp3.py [hunk 20]
```diff
@@ -73,116 +91,86 @@ def sp3_clock_nodata_to_nan(
...
-                9: "category",
-            }
-        )
-        sp3_df.columns = [
-            ["EST", "EST", "EST", "EST", "STD", "STD", "STD", "STD", "P_XYZ", "P_CLK"],
-            ["X", "Y", "Z", "CLK", "X", "Y", "Z", "CLK", "", ""],
-        ]
-        sp3_df.STD = base_xyzc**sp3_df.STD.values
-
+    # Compile the regular expression pattern
```
Comment (treefern) [inline gnssanalysis/gn_io/sp3.py:162]: This comment seems to be orphaned.

File: gnssanalysis/gn_io/sp3.py (hunk 15 above)
Comment (seballgeyer) [reply gnssanalysis/gn_io/sp3.py:99]: Updated, it wwas the original doc, I just did put it in the same sphinx format than all others...

File: .github/workflows/python-cicd-units.yaml (hunk 18 above)
Comment (seballgeyer) [reply .github/workflows/python-cicd-units.yaml:18]: Having the versioner in place, if there is a tag, the version is the tag, else the commit hash. 
We change it in the setup.py so we can install manually some to make some tests.

Comment (seballgeyer) [reply .github/workflows/python-cicd-units.yaml:18]: Oops, this is in the housekeeping branch

**Review by EugeneDu-GA — APPROVED** [review]
Commit `9e96872d` tested working. Looks good for me.

**Review by ronaldmaj — DISMISSED** [review]
With testing complete by @EugeneDu-GA , I'm happy for this to go in as well :+1:

File: gnssanalysis/gn_io/sp3.py [hunk 21]
```diff
@@ -166,11 +166,19 @@ def read_sp3(sp3_path: str, pOnly: bool = True, nodata_to_nan: bool = True) -> _
     if nodata_to_nan:
         sp3_pos_nodata_to_nan(sp3_df)  # Convert 0.000000 (which indicates nodata in the SP3 POS column) to NaN
         sp3_clock_nodata_to_nan(sp3_df)  # Convert 999999* (which indicates nodata in the SP3 CLK column) to NaN
+    print(sp3_df)
```
Comment (ronaldmaj) [inline gnssanalysis/gn_io/sp3.py:169]: This will probably need to be removed before merging in

**Review by ronaldmaj — CHANGES_REQUESTED** [review]

**Comment by EugeneDu-GA** [comment]
[f8e6493](https://github.com/GeoscienceAustralia/gnssanalysis/pull/32/commits/f8e6493aacf3c73e095f60c64b9849d348402636) tested working for sp3 files with velocities, with `pOnly` option either `True` or `False` in `read_sp3()`.

**Review by ronaldmaj — APPROVED** [review]
Print statement removed - happy for this to go in now :+1
